selfdrive/car/hyundai/carcontroller.py

```diff
@@ -1,47 +1,101 @@
+
+  return hud_alert, lane_visible, left_lane_warning, right_lane_warning
 
 class CarController():
   def __init__(self, dbc_name, car_fingerprint):
     self.apply_steer_last = 0
     self.car_fingerprint = car_fingerprint
-    self.lkas11_cnt = 0
-    self.cnt = 0
-    self.last_resume_cnt = 0
     self.packer = CANPacker(dbc_name)
     self.steer_rate_limited = False
+    self.resume_cnt = 0
+    self.last_resume_frame = 0
+    self.last_lead_distance = 0
 
-  def update(self, enabled, CS, actuators, pcm_cancel_cmd, hud_alert):
+  def update(self, enabled, CS, frame, actuators, pcm_cancel_cmd, visual_alert,
+              left_line, right_line, left_lane_depart, right_lane_depart):
 
     ### Steering Torque
     new_steer = actuators.steer * SteerLimitParams.STEER_MAX
     apply_steer = apply_std_steer_torque_limits(new_steer, self.apply_steer_last, CS.steer_torque_driver, SteerLimitParams)
     self.steer_rate_limited = new_steer != apply_steer
 
-    if not enabled:
+    # disable if steer angle reach 90 deg, otherwise mdps fault in some models
+    lkas_active = enabled and abs(CS.angle_steers) < 90.
+    # fix for Genesis hard fault at low speed
+    if CS.v_ego < 16.7 and self.car_fingerprint == CAR.GENESIS:
+      lkas_active = 0
+
+    if not lkas_active:
       apply_steer = 0
 
-    steer_req = 1 if enabled else 0
+    steer_req = 1 if apply_steer else 0
 
     self.apply_steer_last = apply_steer
 
+    hud_alert, lane_visible, left_lane_warning, right_lane_warning =\
+            process_hud_alert(enabled, self.car_fingerprint, visual_alert,
+            left_line, right_line,left_lane_depart, right_lane_depart)
+
     can_sends = []
 
-    self.lkas11_cnt = self.cnt % 0x10
-    self.clu11_cnt = self.cnt % 0x10
+    lkas11_cnt = frame % 0x10
+    clu11_cnt = frame % 0x10
 
-    can_sends.append(create_lkas11(self.packer, self.car_fingerprint, apply_steer, steer_req, self.lkas11_cnt,
-                                   enabled, CS.lkas11, hud_alert, keep_stock=True))
+    can_sends.append(create_lkas11(self.packer, self.car_fingerprint, apply_steer, steer_req, lkas11_cnt, lkas_active,
+                                   CS.lkas11, hud_alert, lane_visible, left_lane_depart, right_lane_depart, keep_stock=True))
 
     if pcm_cancel_cmd:
-      can_sends.append(create_clu11(self.packer, CS.clu11, Buttons.CANCEL))
-    elif CS.stopped and (self.cnt - self.last_resume_cnt) > 5:
-      self.last_resume_cnt = self.cnt
-      can_sends.append(create_clu11(self.packer, CS.clu11, Buttons.RES_ACCEL))
+      can_sends.append(create_clu11(self.packer, CS.clu11, Buttons.CANCEL, clu11_cnt))
+
+    if CS.stopped:
+      # run only first time when the car stopped
+      if self.last_lead_distance == 0:
+        # get the lead distance from the Radar
+        self.last_lead_distance = CS.lead_distance
+        self.resume_cnt = 0
+      # when lead car starts moving, create 6 RES msgs
```

Comment:
Confirming this doesn't actually raise the speed?

Reply:
yes since the condition `if CS.stopped` only true when SCC not active and the driver prompted to press Res button (after 3 sec)
only press resume when lead distance changes, spamming the button all time is annoying as  prompted message flashes with audible alert